core/trino-main/src/main/java/io/trino/util/DateTimeUtils.java

```diff
@@ -407,8 +408,6 @@ public int parseInto(ReadWritablePeriod period, String text, int position, Local
             int bestValidPos = position;
             ReadWritablePeriod bestValidPeriod = null;
 
-            int bestInvalidPos = position;
```

Comment:
why change the `parseInto` method?
I can still get tests to work without these changes. I do see that's it's a "simplification" of the old method - however that should again be it's own commit.

Here's my annotations I added to the (old version) method to understand what it was doing:

```java
private static class OrderedPeriodParser
        implements PeriodParser
{
    private final List<PeriodParser> parsers;

    private OrderedPeriodParser(List<PeriodParser> parsers)
    {
        this.parsers = parsers;
    }

    // Tries each parser in order and returns best (longest) match
    @Override
    public int parseInto(ReadWritablePeriod period, String text, int position, Locale locale)
    {
        // bestValidPos - the farthest position that was successfully parsed
        int bestValidPos = position;
        // bestValidPeriod - holds the ReadWritablePeriod corresponding to the best parse
        ReadWritablePeriod bestValidPeriod = null;

        // bestInvalidPos - the position of the farthest failed parse
        int bestInvalidPos = position;

        // try each parser in order
        for (PeriodParser parser : parsers) {
            ReadWritablePeriod parsedPeriod = new MutablePeriod();
            int parsePos = parser.parseInto(parsedPeriod, text, position, locale);
            // some part of input was parsed
            if (parsePos >= position) {
                // current parser found a longer match than current best parse, update best parse state (bestValidPos + bestValidPeriod)
                if (parsePos > bestValidPos) {
                    bestValidPos = parsePos;
                    bestValidPeriod = parsedPeriod;
                    // we reached end of text, further parsing attempts are useless - this is the best possible parse
                    if (parsePos >= text.length()) {
                        break;
                    }
                }
            }
            // parsing failed
            else if (parsePos < 0) {
                // get failure position
                int failurePos = ~parsePos;
                // this failure matched more of the input than previous failures, update best failed parse state (bestInvalidPos)
                if (failurePos > bestInvalidPos) {
                    bestInvalidPos = failurePos;
                }
            }
        }

        // we've tried all parsers, now we need to set the best valid parse and return it's position or return best failed parse position

        // either a parse was found (>) or a zero-length parse was found (==) - in either case we return the period and position of match
        if (bestValidPos > position || (bestValidPos == position)) {
            // Restore the state to the best valid parse.
            if (bestValidPeriod != null) {
                period.setPeriod(bestValidPeriod);
            }
            return bestValidPos;
        }

        // no valid parses were done, we return negated failed position of longest failed parse as in the contract of PeriodParser
        return ~bestInvalidPos;
    }
}
```

Reply:
The current code doesn't make sense because `if (bestValidPos > position || (bestValidPos == position)) {` is always true and the last `return ~bestInvalidPos;` is unreachable. It's better to return earlier when any parser failed. Sure, it can be a separate commit (or even another PR).

Reply:
For now, I will revert this change as it doesn't relate to this improvement directly. Will consider raising another PR later.

Reply:
Reverted.